File: 2597-the-number-of-beautiful-subsets/2597-the-number-of-beautiful-subsets.cpp
class Solution {
public:
    unordered_map<int, int>mp;
    int ans = 0;
    int solve(vector<int>&nums, int k, int ind, int mask) {
        if(ind == nums.size()){
            return 1;
        }
        bool include = true;
        int ans = 0;
        for(int i=0; i<ind; i++) {
            if(mask&(1<<i)) {
                if(abs(nums[i] - nums[ind]) == k) {
                    include = false;
                    break;
                }
            }
        }
        if(include) ans += solve(nums, k, ind+1, mask|(1<<ind));
        ans += solve(nums, k, ind+1, mask);
        return ans;
    }
    int beautifulSubsets(vector<int>& nums, int k) {
        return solve(nums, k, 0, 0) - 1;
        return ans;
    }
};

File: 2597-the-number-of-beautiful-subsets/2597-the-number-of-beautiful-subsets.java
class Solution {
public:
    unordered_map<int, int>mp;
    int solve(vector<int>&nums, int k, int ind) {
        if(ind == nums.size()) return 1;
        int ans = solve(nums, k, ind+1); // when skip
        if(mp[nums[ind] - k] == 0) {
            mp[nums[ind]]++; // take this nums[i]
            ans += solve(nums, k, ind+1);
            mp[nums[ind]]--; // revert this nums[i]
        }
        return ans;
    }
    int beautifulSubsets(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end());
        return solve(nums, k, 0) - 1;
    }
};